rfcs/0199-basic-config.md

```diff
@@ -0,0 +1,91 @@
+
+The `config` crate would also provide a `MnemosConfig` type that is generic over two types:
+
+* A `KernelConfig` type, which contains all kernel-defined configuration, provided by the `kernel` crate
+* A `PlatformConfig` type, which contains all target- or family-specific configuration, expected to be provided by any of the defined `platforms` that mnemos supports
+
+This crate uses the generics to avoid circular dependencies between the config crate and kernel/platform crates. It is intended that "user facing" applications of the config crate (primarily in platform crates) is done by providing type aliases to "hide" the generics.
+
+[databake]: https://docs.rs/databake/latest/databake/
+
+### Kernel changes
+
+The kernel will provide a single `KernelConfig` type, which is a struct that contains all service and daemon configuration values. We will add functionality to the kernel to take a provided `KernelConfig` type, and automatically launch the configured services.
+
+This capability will allow most platform targets to avoid repeated boilerplate `kernel.initialize(...)` calls at startup.
+
+### Platform crate changes
+
+Platform crates will be expected to provide a `PlatformConfig` type, containing all drivers or daemons that are specific or applicable to this platform (other than the ones defined in the kernel itself). Similar to the kernel, they should also provide a single call that takes a `KernelConfig` and `PlatformConfig`, and automatically launches all kernel and platform services.
+
+One platform may have many targets: For example, the Allwinner D1 platform has three targets currently: the `lichee-rv` dev board, the `mq-pro` dev board, and the `beepy`.
+
+With this system in place, each of these targets would no longer have specific binary targets, and would instead share a single binary target with different configuration files, loaded at compile time.
+
+This configuration file could be set via an environment variable such as `MNEMOS_CONFIG=/path/to/beepy.toml`, which would be loaded by a `build.rs` script, and set by `just`.
+
+In this way, customized targets can be made by creating a new configuration file, rather than an entire cargo project.
+
+## Compile Time vs Runtime Configuration
+
+At the moment, most configuration will be done at compile time, with configuration values being "baked into" the binary.
+
+However, we should ensure that all configuration values also implement `serde` traits, so that platforms that CAN load configuration at runtime (from an SD Card or other external sources) are able to.
+
+Platforms that can load configuration at runtime may need to provide a "two stage" initialization process: using enough of the "baked in" configuration to initialize the SD Card or other external source, then continuing with the initialization process after merging or replacing the baked in data with the loaded data.
+
+## Inspiration
```

Comment:
the `rust-osdev/bootloader` crate, a pure-rust bootloader for x86 systems, has a kind of interesting config design where a common config type is defined in the kernel binary, but shared with the bootloader binary and a build tool that links the two. possibly worth looking at for inspiration: https://github.com/rust-osdev/bootloader#usage